### PR TITLE
Generic CloudConfig Drive

### PR DESCRIPTION
**Pull request by olivierlambert**
WIP. Working on some race conditions.

### CODE REVIEW

**Comment by olivierlambert** [comment]
`_importRawVdi` seems to hang :/

File: src/xapi.js [hunk 1]
```diff
@@ -1403,6 +1464,25 @@ export default class Xapi extends XapiBase {
...
+    const buffer = new Buffer(2097152)
+    const fs = fatfs.createFileSystem(fatfsBuffer(buffer))
+    fs.writeFile('user_data', '#cloud-config\nhostname: ' + vm.nameLabel + '\n')
+    fs.writeFile('meta_data.json', '{ "uuid": "' + vm.uuid + '"}')
+    // Import this tar inside the freshly created VDI
+    await this._importRawVdi(fs, null, vdi)
+    // Attach this VDI to the VM
+    await this.attachVdiToVm(vdi.$ref, vm.$ref)
+
+    return true
```
Comment (julien-f) [inline src/xapi.js]: No need to return `true`.

File: src/xapi.js [hunk 2]
```diff
@@ -1391,7 +1451,8 @@ export default class Xapi extends XapiBase {
     return config.slice(4) // FIXME remove the "True" string on the begining
   }
 
-  async createCloudInitConfigDrive (vmId, srId, config) {
+  // Specifc CoreOS Config Drive
```
Comment (julien-f) [inline src/xapi.js]: Specif**i**c.

File: src/xapi.js [hunk 3]
```diff
@@ -1403,6 +1464,25 @@ export default class Xapi extends XapiBase {
     })
   }
 
+  // Generic Config Drive
+  async createCloudInitConfigDrive (vmId, srId, config) {
+    const vm = this.getObject(vmId)
+    const sr = this.getObject(srId)
+    // First, create a small VDI (2MB) which will become the ConfigDrive
+    const vdi = await this.createVdi(2097152, { name_label: 'XO CloudConfigDrive', name_description: undefined, sr: sr.$ref })
```
Comment (julien-f) [inline src/xapi.js]: Please, put each property on its own line.

File: src/xapi.js [hunk 4]
```diff
@@ -1403,6 +1464,25 @@ export default class Xapi extends XapiBase {
...
+    // Then, generate a tar file with all needed configuration
+    const buffer = new Buffer(2097152)
```
Comment (julien-f) [inline src/xapi.js]: Magic constant?

Use a JS constant `CONFIG_DRIVE_SIZE` and comment the value (e.g. `// 2 mebibytes`).

File: src/xapi.js [hunk 5]
```diff
@@ -1403,6 +1464,25 @@ export default class Xapi extends XapiBase {
...
+    const fs = fatfs.createFileSystem(fatfsBuffer(buffer))
+    fs.writeFile('user_data', '#cloud-config\nhostname: ' + vm.nameLabel + '\n')
+    fs.writeFile('meta_data.json', '{ "uuid": "' + vm.uuid + '"}')
```
Comment (julien-f) [inline src/xapi.js]: Use ES6 template strings instead of concatenation.

File: src/xapi.js [hunk 6]
```diff
@@ -1667,6 +1670,34 @@ export default class Xapi extends XapiBase {
...
+    // Create the user_data file
+    await fs.writeFileAsync('openstack/latest/user_data', config)
+
+    // Transform the buffer into a stream
+    const stream = bufferToStream(buffer)
+    await this.importVdiContent(vdi.$id, stream, {
+      length: null,
+      format: VDI_FORMAT_RAW
+    })
+    await this.attachVdiToVm(vdi.$ref, vm.$ref)
```
Comment (julien-f) [inline src/xapi.js]: You can use the private method `this._createVbd(vm, vdi)` directly.

File: src/xapi.js [hunk 7]
```diff
@@ -1667,6 +1670,34 @@ export default class Xapi extends XapiBase {
...
+
+    return true
```
Comment (julien-f) [inline src/xapi.js]: No need to `return true`.

File: src/xapi.js [hunk 8]
```diff
@@ -1667,6 +1670,34 @@ export default class Xapi extends XapiBase {
     })
   }
 
+  // Generic Config Drive
+  async createCloudInitConfigDrive (vmId, srId, config) {
+    const vm = this.getObject(vmId)
+    const sr = this.getObject(srId)
+    // First, create a small VDI (10MB) which will become the ConfigDrive
```
Comment (julien-f) [inline src/xapi.js]: Feel free to leave blank lines to split different code block.

File: src/xapi.js [hunk 9]
```diff
@@ -1667,6 +1670,34 @@ export default class Xapi extends XapiBase {
...
+    await fs.mkdirAsync('openstack/latest')
+    // Create the meta_data file
+    await fs.writeFileAsync('openstack/latest/meta_data.json', '{\n    "uuid": "' + vm.uuid + '"\n}\n')
+    // Create the user_data file
+    await fs.writeFileAsync('openstack/latest/user_data', config)
+
+    // Transform the buffer into a stream
+    const stream = bufferToStream(buffer)
+    await this.importVdiContent(vdi.$id, stream, {
+      length: null,
```
Comment (julien-f) [inline src/xapi.js]: Not needed.

File: src/xapi.js [hunk 10]
```diff
@@ -927,11 +931,9 @@ export default class Xapi extends XapiBase {
       ) {
         await this.resizeVdi(vdi.$id, size)
       }
-
-      // if another SR is set
+      // if another SR is set, move it there
       if (srId) {
-        // TODO
-        throw new Error('Not implemented')
+        this.moveVdi(vdi.$id, srId)
```
Comment (julien-f) [inline src/xapi.js]: Missing `await`.

File: src/xapi.js [hunk 11]
```diff
@@ -1667,6 +1670,32 @@ export default class Xapi extends XapiBase {
     })
   }
 
+  // Generic Config Drive
+  async createCloudInitConfigDrive (vmId, srId, config) {
+    const vm = this.getObject(vmId)
+    const sr = this.getObject(srId)
+
+    // First, create a small VDI (10MB) which will become the ConfigDrive
+    const vdi = await this.createVdi(10 * 1024 * 1024, { name_label: 'XO CloudConfigDrive', name_description: undefined, sr: sr.$ref })
```
Comment (julien-f) [inline src/xapi.js]: I think it would be cleaner to instantiate the buffer first and then create the VDI with a size of `buffer.length`.

**Pull request merged**